README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

dataaimsr <img src="man/figures/logo.png" width = 180 alt="dataaimsr Logo" align="right" />
===========================================================================================

<!-- badges: start -->

[![](https://badges.ropensci.org/428_status.svg)](https://github.com/ropensci/software-review/issues/428)
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03282/status.svg)](https://doi.org/10.21105/joss.03282)
[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://lifecycle.r-lib.org/articles/stages.html)
[![R build
status](https://github.com/ropensci/dataaimsr/workflows/R-CMD-check/badge.svg)](https://github.com/ropensci/dataaimsr/actions)
[![Codecov test
coverage](https://codecov.io/gh/ropensci/dataaimsr/branch/master/graph/badge.svg)](https://app.codecov.io/gh/ropensci/dataaimsr?branch=master)
![pkgdown](https://github.com/ropensci/dataaimsr/workflows/pkgdown/badge.svg)
[![license](https://img.shields.io/badge/license-MIT%20+%20file%20LICENSE-lightgrey.svg)](https://choosealicense.com/)
[![packageversion](https://img.shields.io/badge/Package%20version-1.1.0-orange.svg)](commits/master)
[![Ask Us Anything
!](https://img.shields.io/badge/Ask%20us-anything-1abc9c.svg)](https://github.com/ropensci/dataaimsr/issues/new)
![Open Source
Love](https://badges.frapsoft.com/os/v2/open-source.svg?v=103)
<!-- badges: end -->

**Barneche DR, Coleman G, Fermor D, Klein E, Robinson T, Smith J,
Sheehan JL, Dowley S, Ditton D, Gunn K, Ericson G, Logan M, Rehbein M**
(2021). dataaimsr: An R Client for the Australian Institute of Marine
Science Data Platform API which provides easy access to AIMS Data
Platform. *Journal of Open Source Software*, **6:** 3282. doi:
[10.21105/joss.03282](https://doi.org/10.21105/joss.03282).

Overview
--------

The Australian Institute of Marine Science (AIMS) has a long tradition
in measuring and monitoring a series of environmental parameters along
the tropical coast of Australia. These parameters include long-term
record of sea surface temperature, wind characteristics, atmospheric
temperature, pressure, chlorophyll-a data, among many others. The AIMS
Data Centre team has recently developed the [AIMS Data Platform
API](https://open-aims.github.io/data-platform/) which is a *REST API*
providing JSON-formatted data to users. `dataaimsr` is an **R package**
written to allow users to communicate with the AIMS Data Platform API
using an API key and a few convenience functions to interrogate and
understand the datasets that are available to download. In doing so, it
allows the user to fully explore these datasets in R in whichever
capacity they want (e.g. data visualisation, statistical analyses, etc).
The package itself contains a `plot` method which allows the user to
plot summaries of the different types of dataset made available by the
API. Below we provide a brief context about the existing
[Datasets](#datasets) that can be explored through `dataaimsr`.

Installation
------------

### Requesting an AIMS Data Platform API Key

**AIMS Data Platform** requires an API Key for data requests, [get a key
here](https://open-AIMS.github.io/data-platform/key-request).

The API Key can be passed to the package functions as an additional
`api_key = "XXXX"` argument. **However**, we strongly encourage users to
maintain their API key as a private locally hidden environment variable
(`AIMS_DATAPLATFORM_API_KEY`) in the `.Renviron` file for automatic
loading at the start of an R session. Please read this
[article](https://CRAN.R-project.org/package=httr/vignettes/secrets.html)
which details why keeping your API private is extremely important.

Users can modify their `.Renviron` file by adding the following line:

    AIMS_DATAPLATFORM_API_KEY=XXXXXXXXXXXXX

The `.Renviron` file is usually stored in each users home directory:

<table>
<colgroup>
<col style="width: 35%" />
<col style="width: 64%" />
</colgroup>
<thead>
<tr class="header">
<th>System</th>
<th>.Renviron file locations</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>MS Windows</td>
<td><code>C:\Users\‹username›\.Renviron</code> or <code>C:\Users\‹username›\Documents\.Renviron</code></td>
</tr>
<tr class="even">
<td>Linux / MacOs</td>
<td><code>/home/‹username›/.Renviron</code></td>
</tr>
</tbody>
</table>

### Package

<table>
<colgroup>
<col style="width: 33%" />
<col style="width: 33%" />
<col style="width: 33%" />
</colgroup>
<thead>
<tr class="header">
<th>Type</th>
<th>Source</th>
<th>Command</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Release</td>
<td>CRAN</td>
<td>Not yet available</td>
</tr>
<tr class="even">
<td>Development</td>
<td>GitHub</td>
<td><code>remotes::install_github("ropensci/dataaimsr")</code></td>
</tr>
<tr class="odd">
<td>Development</td>
<td>rOpenSci</td>
<td><code>install.packages("dataaimsr", repos = "https://dev.ropensci.org")</code></td>
</tr>
</tbody>
</table>

Usage
-----

    # assumes that user already has API key saved to
    # .Renviron
    library(dataaimsr)

    # summarised by series
    # for all sites that contain data
    # within a defined date range
    sdf_b <- aims_data("temp_loggers", api_key = NULL,
                       summary = "summary-by-series",
                       filters = list("from_date" = "2018-01-01",
                                      "thru_date" = "2018-12-31"))

    # downloads weather data from site Yongala
    # within a defined date range
    wdf_a <- aims_data("weather", api_key = NULL,
                       filters = list(site = "Yongala",
                                      from_date = "2018-01-01",
                                      thru_date = "2018-01-02"))

More comprehensive examples about how to navigate `dataaimsr` and
interrogate the datasets can be found on our [online
vignettes](https://ropensci.github.io/dataaimsr/articles/).

Datasets
--------

Currently, there are two AIMS long-term monitoring datasets available to
be downloaded through `dataaimsr`:

### Northern Australia Automated Marine Weather And Oceanographic Stations

Automatic weather stations have been deployed by AIMS since 1980. Most
of the stations are along the Great Barrier Reef (GBR) including the
Torres Strait in North-Eastern Australia but there is also a station in
Darwin and one at Ningaloo Reef in Western Australia. Many of the
stations are located on the reef itself either on poles located in the
reef lagoon or on tourist pontoons or other structures. A list of the
weather stations which have been deployed by AIMS and the period of time
for which data may be available can be found on the
[metadata](https://apps.aims.gov.au/metadata/view/0887cb5b-b443-4e08-a169-038208109466)
webpage. **NB:** Records may not be continuous for the time spans given.

### AIMS Sea Water Temperature Observing System (AIMS Temperature Logger Program)

The data provided here are from a number of sea water temperature
monitoring programs conducted in tropical and subtropical coral reefs
environments around Australia. Data are available from approximately 80
GBR sites, 16 Coral Sea sites, 7 sites in North West Western Australia
(WA), 8 Queensland regional ports, 13 sites in the Solitary Islands, 4
sites in Papua New Guinea and 10 sites in the Cocos (Keeling) Islands.
Data are obtained from in-situ data loggers deployed on the reef.
Temperature instruments sample water temperatures every 5-10 minutes
(typically) and are exchanged and downloaded approximately every 12
months. Temperature loggers on the reef-flat are generally placed just
below Lowest Astronomical Tide level. Reef-slope (or where specified as
Upper reef-slope) generally refers to depths 5–9 m while Deep reef-slope
refers to depths of ~20 m. For more information on the dataset and its
usage, please visit the
[metadata](https://apps.aims.gov.au/metadata/view/4a12a8c0-c573-11dc-b99b-00008a07204e)
webpage.

License
-------

`dataaimsr` is provided by the [Australian Institute of Marine
Science](https://www.aims.gov.au) under the MIT License
([MIT](https://opensource.org/licenses/MIT)).

Code of Conduct
---------------

Please note that this package is released with a [Contributor Code of
Conduct](https://ropensci.org/code-of-conduct/). By contributing to this
project, you agree to abide by its terms.

AIMS R package logos
--------------------

Our R package logos use a watercolour map of Australia, obtained with
the [ggmap](https://CRAN.R-project.org/package=ggmap) R package, which
downloads original map tiles provided by [Stamen
Design](https://stamen.com/), under [CC BY
3.0](https://creativecommons.org/licenses/by/3.0), with data from
[OpenStreetMap](https://www.openstreetmap.org/), under [CC BY
SA](https://creativecommons.org/licenses/by-sa/3.0).


NEWS.md
# dataaimsr 1.1.0

* Added the capacity to download daily aggregated Sea Water Temperature Loggers dataset

# dataaimsr 1.0.3

* Added `aims_*` prefix to all exported functions

* `aims_data`, `aims_filter_values` and `aims_expose_attributes` now take a
string target rather than a string DOI as input argument

* `aims_data` and `aims_filter_values` fail gracefully now if the input
filter parameters are incorrect or there is no internet connection

* `aims_data_doi` is now the non-exported `data_doi`

* Important changes pertaining to `aims_data`:
 - Always returns a data.frame
 - Contains it's own class called `aimsdf` which contains print, summary and plot methods
 - Contains three additional exposed helper functions which allow the user to
 extract metadata/citation/parameter attributes.
 - Example set has been reduced to a minimal amount

* plot method for class `aimsdf` displays either a map or a time series

* improved test coverage

# dataaimsr 1.0.2

* Implemented `summary` datasets for the Temperature Loggers dataset via
`aims_data`

* Implemented `expose_attributes()` to show which filters are accepted
by the different datasets

* restrict `filter_values` to expel info on sites, series and parameters 
only

* Using `parsedate` to standardise date strings and account for time zone

* Created new vignette explaining basic usage of package


R/aims_data.R
#' Request data via the AIMS Data Platform API
#'
#' A function that communicates with the
#' the \href{https://open-aims.github.io/data-platform/}{AIMS Data Platform}
#' via the AIMS Data Platform API
#'
#' @param target A \code{\link[base]{character}} vector of length 1 specifying
#' the dataset. Only \code{weather} or \code{temp_loggers} are currently
#' allowed.
#' @param filters A \code{\link[base]{list}} containing a set of
#' filters for the data query (see Details).
#' @param summary Should summary tables (\code{"summary-by-series"} or
#' \code{"summary-by-deployment"}) or daily aggregated data ("daily") be
#' returned instead of full data (see Details)?
#' @param ... Currently unused. Additional arguments to be passed to
#' non-exported internal functions.
#'
#' @details The AIMS Data Platform R Client provides easy access to
#' data sets for R applications to the
#' \href{https://open-aims.github.io/data-platform/}{AIMS Data Platform API}.
#' The AIMS Data Platform requires an API Key for requests, which can
#' be obtained at this
#' \href{https://open-aims.github.io/data-platform/key-request}{link}.
#' It is preferred that API Keys are not stored in code. We recommend
#' storing the environment variable `AIMS_DATAPLATFORM_API_KEY`
#' permanently under the user's \code{.Renviron} file in order to load
#' the API Key automatically.
#'
#' There are two types of data currently available through the
#' [AIMS Data Platform API](https://open-aims.github.io/data-platform/):
#' [Weather](https://weather.aims.gov.au/#/overview) and
#' [Sea Water Temperature Loggers](https://tinyurl.com/h93mcojk).
#' They are searched internally via unique DOI identifiers.
#' Only one data type at a time can be passed to the argument \code{target}.
#'
#' A list of arguments for \code{filters} can be exposed for both
#' [Weather](https://weather.aims.gov.au/#/overview) and
#' [Sea Water Temperature Loggers](https://weather.aims.gov.au/#/overview)
#' using function \code{\link{aims_expose_attributes}}.
#'
#' Note that at present the user can inspect the range of dates for
#' the temperature loggers data only (see usage of argument \code{summary} in
#' the examples below). For that, the argument \code{summary} must be either
#' the string \code{"summary-by-series"} or \code{"summary-by-deployment"}.
#' In those cases, time filters will be ignored.
#' 
#' Details about available dates for each dataset and time series can be
#' accessed via Metadata on
#' \href{https://open-aims.github.io/data-platform/}{AIMS Data Platform API}.
#' We raise this caveat here because these time boundaries are very important;
#' data are collected at very small time intervals, a window of just a few days
#' can yield very large datasets. The query will return and error
#' if it reaches the system's memory capacity.
#'
#' For that same reason, from version 1.1.0 onwards, we are offering the
#' possibility of downloading a mean daily aggregated version. For that, the
#' user must set \code{summary = "daily"}. In this particular case, query filter
#' will be taken into account.
#'
#' @return \code{aims_data} returns a \code{\link[base]{data.frame}} of class
#' \code{\link{aimsdf}}.
#'
#' If \code{summary %in% c("summary-by-series", "summary-by-deployment")},
#' the output shows the summary information for the target dataset (i.e.
#' weather or temperature loggers)
#' (NB: currently, \code{summary} only works for the temperature logger
#' database). If \code{summary} is *not* passed as an additional argument, then
#' the output contains **raw** monitoring data. If \code{summary = "daily"},
#' then the output contains **mean daily aggregated** monitoring data.
#' The output also contains five attributes (empty strings if
#' \code{summary} is passed as an additional argument):
#' \itemize{
#'    \item{\code{metadata}}{a \href{https://www.doi.org/}{DOI} link
#'          containing the metadata record for the data series.}
#'    \item{\code{citation}}{the citation information for the particular
#'          dataset.}
#'    \item{\code{parameters}}{The measured parameters comprised in the
#'          output.}
#'    \item{\code{type}}{The type of dataset. Either "monitoring" if
#'          \code{summary} is not specified, "monitoring (daily aggregation)" if
#'          \code{summary = "daily"}, or a "summary-by-" otherwise.}
#'    \item{\code{target}}{The input target.}
#' }
#'
#' @author AIMS Datacentre \email{adc@aims.gov.au}
#'
#' @importFrom curl has_internet
#'
#' @seealso \code{\link{aims_citation}}, \code{\link{aims_metadata}},
#' \code{\link{aims_parameters}}
#'
#' @examples
#' \dontrun{
#' library(dataaimsr)
#' # assumes that user already has API key saved to
#' # .Renviron
#'
#' # start downloads:
#' # 1. downloads weather data from
#' # site Yongala
#' # within a defined date range
#' wdf_a <- aims_data("weather", api_key = NULL,
#'                    filters = list(site = "Yongala",
#'                                   from_date = "2018-01-01",
#'                                   thru_date = "2018-01-02"))
#'
#' # 2. downloads weather data from all sites
#' # under series_id 64 from Davies Reef
#' # within a defined date range
#' wdf_b <- aims_data("weather", api_key = NULL,
#'                    filters = list(series_id = 64,
#'                                   from_date = "1991-10-18",
#'                                   thru_date = "1991-10-19"))
#' head(wdf_b)
#' range(wdf_b$time)
#'
#' # 3. downloads weather data from all sites
#' # under series_id 64 from Davies Reef
#' # within defined date AND time range
#' wdf_c <- aims_data("weather", api_key = NULL,
#'                    filters = list(series_id = 64,
#'                                   from_date = "1991-10-18T06:00:00",
#'                                   thru_date = "1991-10-18T12:00:00"))
#' head(wdf_c)
#' range(wdf_c$time)
#'
#' # 4. downloads all parameters from all sites
#' # within a defined date range
#' wdf_d <- aims_data("weather", api_key = NULL,
#'                    filters = list(from_date = "2003-01-01",
#'                                   thru_date = "2003-01-02"))
#' # note that there are multiple sites and series
#' # so in this case, because we did not specify a specific
#' # parameter, series within sites could differ by both
#' # parameter and depth
#' head(wdf_d)
#' unique(wdf_d[, c("site", "series_id", "series")])
#' unique(wdf_d$parameter)
#' range(wdf_d$time)
#'
#' # 5. downloads chlorophyll from all sites
#' # within a defined date range
#' wdf_e <- aims_data("weather", api_key = NULL,
#'                    filters = list(parameter = "Chlorophyll",
#'                                   from_date = "2018-01-01",
#'                                   thru_date = "2018-01-02"))
#' # note again that there are multiple sites and series
#' # however in this case because we did specify a specific
#' # parameter, series within sites differ by depth only
#' head(wdf_e)
#' unique(wdf_e[, c("site", "series_id", "series", "depth")])
#' unique(wdf_e$parameter)
#' range(wdf_e$time)
#'
#' # 6. downloads temperature data
#' # summarised by series
#' sdf_a <- aims_data("temp_loggers", api_key = NULL,
#'                    summary = "summary-by-series")
#' head(sdf_a)
#' dim(sdf_a)
#'
#' # 7. downloads temperature data
#' # summarised by series
#' # for all sites that contain data
#' # within a defined date range
#' sdf_b <- aims_data("temp_loggers", api_key = NULL,
#'                    summary = "summary-by-series",
#'                    filters = list("from_date" = "2018-01-01",
#'                                   "thru_date" = "2018-12-31"))
#' head(sdf_b)
#' dim(sdf_b) # a subset of sdf_a
#'
#' # 8. downloads temperature data
#' # summarised by deployment
#' sdf_c <- aims_data("temp_loggers", api_key = NULL,
#'                    summary = "summary-by-deployment")
#' head(sdf_c)
#' dim(sdf_c)
#'
#' # 9. downloads temperature data
#' # within a defined date range, averaged by day
#' sdf_d <- aims_data("temp_loggers", api_key = NULL, summary = "daily",
#'                    filters = list(series = "DAVFL1",
#'                                   from_date = "2018-01-01",
#'                                   thru_date = "2018-01-10"))
#' # note again that there are multiple sites and series
#' # however in this case because we did specify a specific
#' # parameter, series within sites differ by depth only
#' head(sdf_d)
#' unique(sdf_d[, c("site", "series_id", "series", "depth")])
#' unique(sdf_d$parameter)
#' range(sdf_d$time)
#' }
#'
#' @export
aims_data <- function(target, filters = NULL, summary = NA, ...) {
  # dummy variable to allow testing of network
  network <- as.logical(Sys.getenv("NETWORK_UP", unset = TRUE))
  if (!has_internet() | !network) {
    message("It seems that your internet connection is down. You need an ",
            "internet connection to successfully download AIMS data.")
    return(invisible())
  }
  doi <- data_doi(target = target)
  w_doi <- data_doi(target = "weather")
  allowed <- aims_expose_attributes(target = target)
  add_args <- list(...)
  if (!is.na(summary)) {
    if (doi == w_doi) {
      message("Argument \"summary\" is currently only available",
              " for the temperature logger (\"temp_loggers\") dataset.\n",
              " Ignoring \"summary\" entry. See details in",
              " ?aims_expose_attributes")
      summary <- NA
    }
    if (!all(summary %in% allowed$summary)) {
      wrong_s <- setdiff(summary, allowed$summary)
      stop("summary string \"", paste(wrong_s, sep = "; "),
           "\" not allowed; please check ?aims_expose_attributes")
    }
  }
  if (!is.null(filters)) {
    if (!all(names(filters) %in% allowed$filters)) {
      wrong_f <- setdiff(names(filters), allowed$filters)
      stop("filter string \"", paste(wrong_f, sep = "; "),
           "\" not allowed; please check ?aims_expose_attributes")
    }
  }
  all_args <- c(doi = doi, filters = list(filters), summary = summary, add_args)
  results <- do.call(page_data, all_args)
  message(results$links)
  next_url <- results$links$next_page
  more_data <- TRUE
  while (more_data) {
    tryCatch({
      next_res <- next_page_data(next_url, ..., doi = doi)
      results$data <- rbind(results$data, next_res$data)
      if ("links" %in% names(next_res) &&
          "next_page" %in% names(next_res$links)) {
        message(next_res$links)
        next_url <- next_res$links$next_page
      } else {
        more_data <- FALSE
      }
    }, warning = function(war) {
      more_data <<- FALSE
    }, error = function(err) {
      more_data <<- FALSE
    })
    message(paste("Result count:", nrow(results$data)))
  }
  final_wrap <- "full_data"
  data_type <- "monitoring"
  if (!is.na(summary)) {
    if (summary == "daily") {
      data_type <- "monitoring (daily aggregation)"
    } else {
      final_wrap <- "summary_data"
    }
  }
  if (final_wrap == "summary_data") {
    attr(results, "citation") <- ""
    attr(results, "metadata") <- ""
    attr(results, "parameters") <- ""
    attr(results, "type") <- summary
    attr(results, "target") <- target
  } else {
    if (!inherits(results$data, "data.frame") |
          (inherits(results$data, "data.frame") && nrow(results$data) == 0)) {
      message("Failed to download monitoring data; Did you specify ",
              "appropriate filter values for your chosen filters?")
      return(invisible())
    }
    attr(results$data, "citation") <- results$citation
    attr(results$data, "metadata") <- results$metadata
    attr(results$data, "parameters") <- unique(results$data$parameter)
    attr(results$data, "type") <- data_type
    attr(results$data, "target") <- target
    results <- results$data
  }
  class(results) <- c("aimsdf", class(results))
  results
}
